### PR TITLE
Allow excluding windows from being closed via Escape key

### PR DESCRIPTION
**Pull request by falbrechtskirchinger**
Add a "pin" function to in-game windows. The function replaces the shade (aka. minimize) button and the window can instead be shaded by double-clicking the title bar.
Pinned windows are excluded from being closed by the window manager when hitting the Escape key.

The setting `interface.enable_window_pinning` has to be set to `1` for the new behavior to take effect.

An option can be added to the UI once the options screen has been reorganized.

To-do:
- [x] `Alt-W` should still close all windows.
- [x] Right-clicking also closes pinned windows. Should it? I'm leaning toward "No" at the moment.
- [x] Save and restore the pinned state as part of the window settings.
- [x] ~~The feature has poor discoverability. Should it be added as a separate button instead?~~
- [x] Rebase after #1608 and #1609.
- [x] Add unit tests?

### CODE REVIEW

**Comment by Flow86** [comment]
> The feature has poor discoverability. Should it be added as a separate button instead?

adding a small "pin" button left of the minimize button of windows would be great.

**Comment by Spikeone** [comment]
Although one should be able to disable the pin button (for people who like the original interface). Will think about a pin icon.

**Comment by falbrechtskirchinger** [comment]
FYI, this is what the button looks like at the moment:

pin_window_button

(A reddened hover state of the corner button.)

**Comment by falbrechtskirchinger** [comment]
I've updated the PR description with the new mechanic, developed after some back-and-forth with @Spikeone on Discord.

To summarize the rationale: An additional button would've been difficult to fit on some windows (e.g., HQ) and required new assets for the top part of the window chrome. Once we add a setting to the options screen – with a tooltip explaining the behavior – this should be discoverable enough.

The pin/unpin icons are a bit difficult to tell apart (at least to my eyes and without #1594 :wink:) and @Spikeone created another set of icons using a lock, which is visually easier to distinguish. We could rename the feature and change the icons as well.

@Flow86 Thoughts?

Otherwise, work is paused until after #1608 and #1609 have been (ideally) merged.

**Comment by falbrechtskirchinger** [comment]
For easier viewing, these are all the new assets at 400%:
resource_new


<details>
  <summary>Commands used for future reference:</summary>

```sh
scale=4
pad=4
s=$((16*scale + pad*2))
for f in *.bmp; do magick "$f" -filter point -resize $((scale*100))% -background white -gravity center \
                               -extent ${s}x${s} "${f%.bmp}.png"; done
magick montage -background white -tile 2x3 000*.png 003*.png 001*.png 004*.png 002*.png 005*.png
               -mode Concatenate resource_new.png
```
</details>

File: libs/s25main/Loader.cpp
Comment (Flow86) [inline libs/s25main/Loader.cpp:218]: please don't call it `new` ("its new until its old") - perhaps use `resource_rttr` instead?

Comment (Flamefire) [reply libs/s25main/Loader.cpp:218]: We have `io_new` already... Maybe simply put the new icons into that folder?

Comment (falbrechtskirchinger) [reply libs/s25main/Loader.cpp:218]: I don't really care either way. Just tell me where to put it … \*pause for inadvertent joke\*

Comment (Flamefire) [reply libs/s25main/Loader.cpp:218]: Ok, then I make the call and say `io_new` as that contains other icons as well.

Comment (falbrechtskirchinger) [reply libs/s25main/Loader.cpp:218]: Alright. FYI, with regards to naming, I followed the original in case that seems odd.

E.g., `closicdn`/`closicsel`/`closicup` = Close Icon Down/Selected/Up

Comment (Flamefire) [reply libs/s25main/Loader.cpp:218]: Don't mind that but doesn't matter much as we currently refer to images by index rather than by name. Hence we need names such as `014.nx14.ny0.bmp` where it encodes the index and the (draw-)origin

Comment (falbrechtskirchinger) [reply libs/s25main/Loader.cpp:218]: Yeah, I could have made use of the origin, but exporting full-size images was easier.

**Comment by falbrechtskirchinger** [comment]
I want to add a new category called `[tweakables]` to the settings. Intended for all the little knobs like `enable_window_pinning` etc. This would be useful for #1604 as well.
Thoughts anyone?

**Comment by Flamefire** [comment]
> I want to add a new category called `[tweakables]` to the settings. Intended for all the little knobs like `enable_window_pinning` etc. This would be useful for #1604 as well.

What would you put there from the other PR? IMO the interface category is fitting, i.e. similar to revert-mouse

**Comment by falbrechtskirchinger** [comment]
> > I want to add a new category called `[tweakables]` to the settings. 
>
> What would you put there from the other PR? IMO the interface category is fitting, i.e. similar to revert-mouse

One or two settings for the "Follow object" changes. @Spikeon was arguing for some control to retain the vanilla experience.

I'm fine with those going into `[interface]` but was thinking ahead a little. Might be nice to have particularly granular stuff in one place. Just a thought.

**Comment by spikeon** [comment]
I believe you meant to tag @Spikeone i am unfamiliar with this repo

**Comment by falbrechtskirchinger** [comment]
> I believe you meant to tag @Spikeone i am unfamiliar with this repo

My apologies! You're correct.

**Comment by Spikeone** [comment]
> What would you put there from the other PR? IMO the interface category is fitting, i.e. similar to revert-mouse

I think the idea to put multiple settings into that screen, maybe some that don't change the interface but other behaviors (I have no idea at the moment what settings could be client side only and change behavior, and are not interface options). Maybe at some point we'd split tweakables and interface again.

**Comment by falbrechtskirchinger** [comment]
Oh, the original purpose of the `[tweakables]` category came back to me. I wanted to suggest putting some of the things that are now hardcoded constants there. For example, the (not yet implemented) pick radius in pixels for #1604.
All these values would also be perfectly at home in the `[interface]` category, I just thought it might be useful to separate out the smaller things, before `[interface]` gets crowded. :man_shrugging:

File: libs/s25main/ingameWindows/IngameWindow.cpp
Comment (falbrechtskirchinger) [inline libs/s25main/ingameWindows/IngameWindow.cpp:326]: This seems a bit silly. Should just pre-calculate the arrays.

File: libs/s25main/Settings.h
Comment (falbrechtskirchinger) [inline libs/s25main/Settings.h:108]: I want to rename this to just `windowPinning`. Shorter and leaves the door open if `true`/`false` aren't enough. (For example, to add a third value for closing on right-click.)

**Review by falbrechtskirchinger — COMMENTED** [review]
Besides two minor things and a final decision on `[tweakables]`, this seems ready to me.

File: libs/s25main/Settings.h
Comment (falbrechtskirchinger) [reply libs/s25main/Settings.h:108]: On second thought, doesn't make much of a difference. The current setting names are a mix between imperative mood, past participles, and noun phrases … it's not going to be more consistent one way or the other.

**Review by Flamefire — APPROVED** [review]

File: libs/s25main/Settings.h
Comment (Flamefire) [reply libs/s25main/Settings.h:108]: Yes I know it's a bit "grown historically"...

**Review by Flamefire — APPROVED** [review]

**Comment by falbrechtskirchinger** [comment]
The required test runs for the auto-merge seem to have been canceled so I rebased onto master and force-pushed over your merge commit to trigger another run. There's one cosmetic change I forgot to push previously, otherwise, it's unchanged.

**Review by Flamefire — APPROVED** [review]

**Comment by Flamefire** [comment]
The appveyor build failed due to a random issue unrelated to this. I added a commit to fix this to your branch so you don't need to do anything.

**Pull request merged**
